### PR TITLE
Algolia search

### PR DESCRIPTION
**Pull request by mikekeehnen**
### Summary
🔎🧮🚀

This pull request adds a new package `@graphcommerce/algolia-search` that implements Algolia search functionality for the GraphCommerce framework. It defines various components, hooks, types, and plugins to integrate Algolia search with the existing product list and filter components from the `@graphcommerce/magento-product` package. It also updates the documentation and the example project to reflect the new feature.

> _Oh we're the crew of the GraphCommerce ship_
> _And we sail the web with Algolia_
> _We've got hooks and chips and plugins galore_
> _To make our search and filter glow_

### CODE REVIEW

File: docs/framework/config.md [hunk 1]
```diff
@@ -189,7 +217,7 @@ Hide the wishlist functionality for guests.
 
 #### `wishlistIgnoreProductWishlistStatus: Boolean`
 
-Ignores whether a product is already in the wishlist, makes the toggle an add only.
+Ignores wether a product is already in the wishlist, makes the toggle an add only.
```
Comment (timhofman) [inline docs/framework/config.md:220]: a 'wether' as a castrated ram

File: packages/algolia-search/hooks/useAlgoliaResults.ts [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+import { useHits } from 'react-instantsearch-hooks'
+import { AlgoliaHit } from '../lib/types'
+
+function hitsToProduct(items: AlgoliaHit[], currency?: string | null) {
+  const mapHits = items.map((item) => {
+    const currentCurrency = (currency ?? Object.keys(item.price)[0]) as CurrencyEnum
+    const price = item.price[currentCurrency]
+    const productUrlSplit = item.url.split('/')
+    const productUrl = productUrlSplit[productUrlSplit.length - 1]
+    const url_key = productUrl.substring(0, productUrl.length - 5)
```
Comment (paales) [inline packages/algolia-search/hooks/useAlgoliaResults.ts:14]: This should take the product_url_suffix in consideration. Can gotten from the storeconfig

File: packages/algolia-search/hooks/useSearchRoute.ts [hunk 3]
```diff
@@ -0,0 +1,6 @@
+import { useRouter } from 'next/router'
+
+export function useSearchRoute() {
+  const router = useRouter()
+  return router.asPath.includes('/search')
+}
```
Comment (paales) [inline packages/algolia-search/hooks/useSearchRoute.ts:6]: Should we move this to the magento-search package?

File: packages/algolia-search/Config.graphqls [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+
+  """
+  Configure your Algolia Search index
+  """
+  algoliaSearchIndex: String
+
+  """
+  Configures Algolia filterable attributes
+  """
+  algoliaFilterAttributes: [AlgoliaFilterAttribute]
```
Comment (paales) [inline packages/algolia-search/Config.graphqls:35]:
```suggestion
  algoliaFilterAttributes: [AlgoliaFilterAttribute!]
```

File: packages/algolia-search/lib/configuration.ts [hunk 5]
```diff
@@ -0,0 +1,3 @@
+export const applicationId = import.meta.graphCommerce.algoliaApplicationId ?? ''
+export const searchOnlyApiKey = import.meta.graphCommerce.algoliaSearchOnlyApiKey ?? ''
+export const searchIndex = import.meta.graphCommerce.algoliaSearchIndex ?? ''
```
Comment (paales) [inline packages/algolia-search/lib/configuration.ts:3]: If we make the fields required:

```suggestion
export const applicationId = import.meta.graphCommerce.algoliaApplicationId
export const searchOnlyApiKey = import.meta.graphCommerce.algoliaSearchOnlyApiKey
export const searchIndex = import.meta.graphCommerce.algoliaSearchIndex
```

Comment (paales) [reply packages/algolia-search/lib/configuration.ts:3]: Also, this doesn't take the storefront override in consideration

File: packages/algolia-search/plugins/AlgoliaFiltersPlugin.tsx [hunk 6]
```diff
@@ -0,0 +1,22 @@
...
+  if (!search) return <Prev {...props} />
+
+  const AlgoliaFilters = dynamic(
+    async () => (await import('../components/Filters/AlgoliaFilters')).AlgoliaFilters,
+  )
+
+  return <AlgoliaFilters {...rest} />
+}
+
+export const Plugin = AlgoliaFiltersPlugin
```
Comment (paales) [inline packages/algolia-search/plugins/AlgoliaFiltersPlugin.tsx:22]: Maybe the bundle size increate is inevitable? Lets discuss

File: packages/algolia-search/plugins/AlgoliaSearchResultsPlugin.tsx [hunk 7]
```diff
@@ -0,0 +1,24 @@
...
+  if (!search) return <Prev {...props} />
+
+  const AlgoliaResults = dynamic(
+    async () => (await import('../components/Filters/AlgoliaResults')).AlgoliaResults,
+  )
+
+  return <AlgoliaResults {...props} />
+}
+
+export const Plugin = AlgoliaSearchPlugin
```
Comment (paales) [inline packages/algolia-search/plugins/AlgoliaSearchResultsPlugin.tsx:24]: This means that we completely switch to client side rendering, am I right? Is there a way to also do SSR for this? Maybe execute the search in getStaticProps as well?

File: packages/algolia-search/plugins/AlgoliaSearchContextPlugin.tsx [hunk 8]
```diff
@@ -0,0 +1,23 @@
...
+import algoliasearch from 'algoliasearch/lite'
+import { InstantSearch } from 'react-instantsearch-hooks'
+import { applicationId, searchOnlyApiKey } from '../lib/configuration'
+
+export const component = 'SearchContext'
+export const exported = '@graphcommerce/magento-search'
+export const ifConfig: IfConfig = 'demoMode'
+
+const searchClient = algoliasearch(applicationId, searchOnlyApiKey)
+const searchIndex = import.meta.graphCommerce.algoliaSearchIndex ?? ''
```
Comment (paales) [inline packages/algolia-search/plugins/AlgoliaSearchContextPlugin.tsx:12]: Also exported from configuration

File: packages/algolia-search/lib/configuration.ts (hunk 5 above)
Comment (paales) [reply packages/algolia-search/lib/configuration.ts:3]: Might it be an idea to create a useAlgoliaConfiguration() hook which also allows us to use the useStoreconfig hook so we can override the configuration for a specific storefront?

**Review by paales — CHANGES_REQUESTED** [review]

File: packages/algolia-search/plugins/AlgoliaSearchResultsPlugin.tsx (hunk 7 above)
Comment (mikekeehnen) [reply packages/algolia-search/plugins/AlgoliaSearchResultsPlugin.tsx:24]: The `react-instantsearch-hooks` library does support SSR, although it looks like we won't be able to implement this through plugins. 

https://www.algolia.com/doc/guides/building-search-ui/going-further/server-side-rendering

Comment (mikekeehnen) [reply packages/algolia-search/plugins/AlgoliaSearchResultsPlugin.tsx:16]: We could add the SSR support and include some extra plugin installation steps to activate this feature. Adding this to the README!

File: packages/algolia-search/lib/configuration.ts (hunk 5 above)
Comment (mikekeehnen) [reply packages/algolia-search/lib/configuration.ts]: De plugin supports storefront specific search indexes and configurations, but the application id is app based. Do you think we need to make this storefront specific as well?

**Pull request merged**